kornia/geometry/transform/thin_plate_spline.py

```diff
@@ -0,0 +1,148 @@
+from typing import Tuple
+
+import torch
+import torch.nn as nn
+
+from kornia.utils import _extract_device_dtype
+
+__all__ = [
+    "get_tps_parameters",
+    "warp_points_tps",
+    "warp_img_tensor_tps"
+]
+
+# utilities for computing thin plate spline transforms
+
+
+def _pair_square_euclidean(tensor1: torch.Tensor, tensor2: torch.Tensor) -> torch.Tensor:
+    """Compute the pairwise squared euclidean distance matrices (B, N, M) between two tensors
+        with shapes (B, N, C) and (B, M, C)."""
+    # ||t1-t2||^2 = (t1-t2)^T(t1-t2) = t1^T*t1 + t2^T*t2 - 2*t1^T*t2
+    t1_sq: torch.Tensor = tensor1.mul(tensor1).sum(dim=-1, keepdim=True)
+    t2_sq: torch.Tensor = tensor2.mul(tensor2).sum(dim=-1, keepdim=True).transpose(1, 2)
+    t1_t2: torch.Tensor = tensor1.matmul(tensor2.transpose(1, 2))
+    square_dist: torch.Tensor = -2 * t1_t2 + t1_sq + t2_sq
+    square_dist: torch.Tensor = square_dist.clamp(min=0)  # handle possible numerical errors
+    return square_dist
+
+
+def _kernel_distance(squared_distances: torch.Tensor, eps: float = 1e-8) -> torch.Tensor:
+    """Compute the TPS kernel distance function: r^2*log(r), where r is the euclidean distance.
+    Since log(r) = 1/2*log(r^2), this function takes the squared distance matrix and calculates
+    0.5 * r^2 * log(r^2)."""
+    # r^2 * log(r) = 1/2 * r^2 * log(r^2)
+    return 0.5 * squared_distances * squared_distances.add(eps).log()
+
+
+def get_tps_parameters(points_src: torch.Tensor, points_target: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Compute the TPS transform parameters that warp source points to target points.
+
+    The input to this function is a tensor of (x, y) source points (B, N, 2) and a corresponding tensor of target
+    (x, y) points (B, N, 2).
+
+    The return value is a tuple containing two weight tensors. The first weight tensor contains the kernel weights
+    (B, N, 2); the second tensor contains the affine weights (B, 3, 2). In both cases, the last dimension contains
+    the weights for the x-transform and y-transform respectively.
+
+    Args:
+        points_src (torch.Tensor): batch of source points (B, N, 2) as (x, y) coordinate vectors
+        points_target (torch.Tensor): batch of target points (B, N, 2) as (x, y) coordinate vectors
+
+    Returns:
+        Tuple[torch.Tensor, torch.Tensor]: the (B, N, 2) kernel weights and (B, 3, 2) affine weights
+    """
```

Comment:
for the docstring optional to add :math:`(B,N,2)` so that is rendered in latex style.